File: backend/app/schemas/venue.py
from pydantic import BaseModel, ConfigDict, Field, field_serializer
import uuid


class ZoneResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    venue_id: uuid.UUID
    name: str
    zone_type: str
    capacity: int
    current_count: int
    current_density: float
    density_level: str
    lat_center: float
    lng_center: float
    polygon: list[dict[str, float]] | None
    is_open: bool
    is_exit_blocked: bool

    @field_serializer("zone_type", "density_level")
    def serialize_enum_fields(self, value: str) -> str:
        return str(value.value if hasattr(value, "value") else value).lower()


class VenueResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    name: str
    address: str
    city: str
    country: str
    capacity: int
    lat_center: float
    lng_center: float
    floor_plan_url: str | None = None
    floor_plan_bounds: dict[str, object] | None = None
    timezone: str
    is_active: bool


class VenueDetailResponse(VenueResponse):
    """Full venue response including all zones. Used for initial app bootstrap."""

    zones: list[ZoneResponse] = Field(default_factory=list)


class RouteStep(BaseModel):
    instruction: str
    distance: str | float
    duration: str | float
    start_lat: str | float
    start_lng: str | float


class RouteResponse(BaseModel):
    """Navigation route from Google Maps Directions API."""

    distance: str
    duration: str
    steps: list[RouteStep]
    polyline: str


class ZoneUpdateRequest(BaseModel):
    """Fields to update on a Zone."""

    is_open: bool | None = None
    is_exit_blocked: bool | None = None

File: frontend/src/types/venue.ts
export type ZoneType =
  | 'entry'
  | 'concourse'
  | 'seating'
  | 'food'
  | 'restroom'
  | 'medical'
  | 'exit'
  | 'staff';

export interface PolygonPoint {
  readonly lat: number;
  readonly lng: number;
}

export interface Zone {
  readonly id: string;
  readonly venueId: string;
  readonly venue_id?: string;
  readonly name: string;
  readonly zoneType: ZoneType;
  readonly zone_type?: ZoneType;
  readonly capacity: number;
  readonly currentCount: number;
  readonly current_count?: number;
  readonly currentDensity: number;
  readonly current_density?: number;
  readonly densityLevel: import('./crowd').DensityLevel;
  readonly density_level?: import('./crowd').DensityLevel;
  readonly latCenter: number;
  readonly lat_center?: number;
  readonly lngCenter: number;
  readonly lng_center?: number;
  readonly polygon: PolygonPoint[] | null;
  readonly isOpen: boolean;
  readonly is_open?: boolean;
  readonly isExitBlocked: boolean;
  readonly is_exit_blocked?: boolean;
}

export interface FloorPlanBounds {
  readonly north: number;
  readonly south: number;
  readonly east: number;
  readonly west: number;
}

export interface Venue {
  readonly id: string;
  readonly slug?: string | null;
  readonly name: string;
  readonly address: string;
  readonly city: string;
  readonly country: string;
  readonly capacity: number;
  readonly latCenter: number;
  readonly lat_center?: number;
  readonly lngCenter: number;
  readonly lng_center?: number;
  readonly floorPlanUrl: string | null;
  readonly floor_plan_url?: string | null;
  readonly floorPlanBounds: FloorPlanBounds | null;
  readonly floor_plan_bounds?: FloorPlanBounds | null;
  readonly timezone: string;
  readonly isActive: boolean;
  readonly is_active?: boolean;
  readonly zones: Zone[];
}

export interface RouteResponse {
  distance: string;
  duration: string;
  steps: Array<{
    instruction: string;
    distance: string;
    duration: string;
    start_lat: number | string;
    start_lng: number | string;
  }>;
  polyline: string;
}
